scanner/src/funTest/kotlin/experimental/DefaultNestedProvenanceResolverFunTest.kt

```diff
@@ -0,0 +1,202 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.scanner.experimental
+
+import io.kotest.assertions.throwables.shouldThrow
+import io.kotest.core.spec.style.WordSpec
+import io.kotest.matchers.should
+import io.kotest.matchers.shouldBe
+
+import java.io.IOException
+
+import org.ossreviewtoolkit.model.ArtifactProvenance
+import org.ossreviewtoolkit.model.Hash
+import org.ossreviewtoolkit.model.RemoteArtifact
+import org.ossreviewtoolkit.model.RepositoryProvenance
+import org.ossreviewtoolkit.model.VcsInfo
+import org.ossreviewtoolkit.model.VcsType
+import org.ossreviewtoolkit.utils.Os
+import org.ossreviewtoolkit.utils.test.containExactly
+
+class DefaultNestedProvenanceResolverFunTest : WordSpec() {
+    private val resolver = DefaultNestedProvenanceResolver()
+
+    init {
+        "Resolving an artifact provenance" should {
+            "Return only the artifact provenance" {
+                val provenance = ArtifactProvenance(
+                    sourceArtifact = RemoteArtifact(
+                        url = "",
+                        hash = Hash.NONE
+                    )
+                )
+
+                resolver.resolveNestedProvenance(provenance) shouldBe
+                        NestedProvenance(root = provenance, subRepositories = emptyMap())
+            }
+        }
+
+        "Resolving a repository provenance" should {
+            "Return the correct root provenance" {
+                val provenance = RepositoryProvenance(
+                    vcsInfo = VcsInfo(
+                        type = VcsType.GIT,
+                        url = "https://github.com/oss-review-toolkit/ort-governance.git",
+                        revision = "1440358978888bfc4b6fbf356e4e6be005c9a25e"
+                    ),
+                    resolvedRevision = "1440358978888bfc4b6fbf356e4e6be005c9a25e"
+                )
+
+                resolver.resolveNestedProvenance(provenance) shouldBe
+                        NestedProvenance(root = provenance, subRepositories = emptyMap())
+            }
+
+            "Find recursive Git submodules" {
+                val provenance = RepositoryProvenance(
+                    vcsInfo = VcsInfo(
+                        type = VcsType.GIT,
+                        url = "https://github.com/oss-review-toolkit/ort-test-data-git-submodules.git",
+                        revision = "fcea94bab5835172e826afddb9f6427274c983b9"
+                    ),
+                    resolvedRevision = "fcea94bab5835172e826afddb9f6427274c983b9"
+                )
+
+                val tree = resolver.resolveNestedProvenance(provenance)
```

Comment:
This is still called "tree", not sure if it's intentional. Could be e.g. just "repo".